### PR TITLE
Add data hashing functions

### LINKED ISSUE #271: Expose additional Cryptographic Hashing APIs
As with script hashing (#245)  there's some cryptographic functions which are necessary in many cases to construct or test scripts, which Plutus makes available only in `plutus-apps`. 

- `pubKeyHash` , which uses `Blake2b_224`
- `datumHash`, which uses `Blake2b_256`

In PlutusTx, we also have:
- `Sha2_256` & `Sha3_256`, which have builtins in PlutusTx. 

These aren't used for anything fundamental (afaik), but would be useful to expose APIs for incase they are relevant.

### PR DESCRIPTION
**Pull request by TotallyNotChase**
Resolves #271

### CODE REVIEW

**Comment by TotallyNotChase** [comment]
This is missing `pubKeyHash` and that should be it I believe. However, The `PubKey` export seems to have been removed from `plutus-ledger-api`....

**Review by L-as — COMMENTED** [review]
Do we want to tie this in with `pserialiseData` somehow?

File: Plutarch/Api/V1.hs [hunk 1]
```diff
@@ -139,3 +143,15 @@ mintingPolicySymbol = coerce scriptHash
 -- | Hash a StakeValidator, with the correct prefix for Plutus V1
 stakeValidatorHash :: Plutus.StakeValidator -> Plutus.StakeValidatorHash
 stakeValidatorHash = coerce scriptHash
+
+-- | Hash a Datum.
+datumHash :: Plutus.Datum -> Plutus.DatumHash
+datumHash = coerce . dataHash
```
[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply Plutarch/Api/V1.hs:149]: I think I have to do `coerce $ dataHash @Plutus.Datum` to make it work, `coerce dataHash` raises an ambiguous type variable error preventing the `ToData` constraint to be solved (which is really weird).

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Api/V1.hs:149]: You should just give it a type rather than using type application.

**Comment by L-as** [comment]
New UPLC built-in

File: Plutarch/Api/V1.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply Plutarch/Api/V1.hs:149]: By "give it a type", do you mean using something like `coerce (dataHash :: Plutus.Datum -> Plutus.BuiltinByteString)`, or something else?

**Comment by TotallyNotChase** [comment]
If it's a UPLC built in, we should add it separately. We could probably use it in `dataHash` and `plift` the result, but I like the pure haskell approach better

**Review by L-as — APPROVED** [review]
LGTM

File: Plutarch/Api/V1.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Api/V1.hs:149]: I meant `coerce (dataHash :: t)`, it's not really important, but it's theoretically more performant. `coerce . dataHash` turns into `id . dataHash` really.

**Pull request merged**